ouroboros-consensus/src/consensus-testlib/Test/Ouroboros/Consensus/ChainGenerator/Adversarial.hs

```diff
@@ -0,0 +1,521 @@
+  , NoSuchAdversarialChainSchema (NoSuchAdversarialBlock, NoSuchCompetitor, NoSuchIntersection)
+  , SomeCheckedAdversarialRecipe (SomeCheckedAdversarialRecipe)
+  , checkAdversarialRecipe
+  , uniformAdversarialChain
+    -- * Testing
+  , AdversarialViolation (BadAnchor, BadCount, BadRace)
+  , AnchorViolation (HonestActiveMustAnchorAdversarial)
+  , ChainSchema (ChainSchema)
+  , RaceViolation (AdversaryWonRace, rvAdv, rvHon)
+  , checkAdversarialChain
+  ) where
+
+import           Control.Applicative ((<|>))
+import           Control.Monad (void, when)
+import qualified Control.Monad.Except as Exn
+import           Data.Proxy (Proxy (Proxy))
+import qualified System.Random.Stateful as R
+import qualified Test.Ouroboros.Consensus.ChainGenerator.BitVector as BV
+import qualified Test.Ouroboros.Consensus.ChainGenerator.Counting as C
+import           Test.Ouroboros.Consensus.ChainGenerator.Honest
+                     (ChainSchema (ChainSchema))
+import           Test.Ouroboros.Consensus.ChainGenerator.Params (Asc,
+                     Delta (Delta), Kcp (Kcp), Scg (Scg))
+import qualified Test.Ouroboros.Consensus.ChainGenerator.RaceIterator as RI
+import qualified Test.Ouroboros.Consensus.ChainGenerator.Slot as S
+import           Test.Ouroboros.Consensus.ChainGenerator.Slot
+                     (E (ActiveSlotE, EmptySlotE, SlotE))
+import qualified Test.Ouroboros.Consensus.ChainGenerator.Some as Some
+
+-----
+
+data AnchorViolation =
+    -- | An honest active slot must immediately precede the adversarial interval
+    HonestActiveMustAnchorAdversarial
+  |
+    -- | The were not exactly 'arPrefix' many active slots preceding @adv@
+    WrongNumberOfHonestPredecessors
+  deriving (Eq, Read, Show)
+
+-- | A violation of Adversarial Chains in Stability Windows
+--
+-- INVARIANT: @'C.windowLast' 'rvAdv' < 'C.windowLast' 'rvHon' + 'Delta'@
+--
+-- INVARIANT: @'C.windowStart' 'rvHon' <= 'C.windowStart' 'rvAdv'@
+data RaceViolation hon adv = AdversaryWonRace {
+    -- | The adversarial race window
+    rvAdv :: !(RI.Race adv)
+  ,
+    -- | The honest race window
+    rvHon :: !(RI.Race hon)
+  }
+  deriving (Eq, Read, Show)
+
+data AdversarialViolation hon adv =
+    BadAnchor !AnchorViolation
+  |
+    BadCount
+  |
+    BadRace !(RaceViolation hon adv)
+  deriving (Eq, Read, Show)
+
+-- | Check the chain matches the given 'AdversarialRecipe', especially that it
+-- satisfies Adversarial Chains in Stability Windows and Length of Competing
+-- Chains.
+--
+-- Definition of a /Praos Race Window/ of a chain. It is an interval of slots
+-- that contains at least @k+1@ blocks of the chain and exactly 'Delta' slots
+-- after the @k+1@st block.
+--
+-- Definition of the /Length of Competing Chains/. We assume every adversarial
+-- chain contains at most @k@ blocks in the Praos Race Window anchored at the
+-- the intersection.
+--
+-- Definition of the /Adversarial Chains in Stability Windows/.
```

Comment:
Maybe "Races in Stability Windows" is shorter and more evocative.

Reply:
"Definition of _the_ /.../" would require a noun phrase to be the ellipses, such as Assumption, Property, etc.

Reply:
Got it.

Reply:
Adopted in 491a9d9